spec/ruby/core/module/name_spec.rb

```diff
@@ -6,6 +6,40 @@
     Module.new.name.should be_nil
   end
 
+  ruby_version_is "3.3" do
+    it "can assign a temporary name" do
+      m = Module.new
+      m.name.should be_nil
+
+      m.set_temporary_name("Foo")
+      m.name.should == "Foo"
+
+      m.set_temporary_name(nil)
+      m.name.should be_nil
+    end
+
+    it "can't assign empty string as name" do
+      m = Module.new
+      -> { m.set_temporary_name("") }.should raise_error(ArgumentError, "empty class/module name")
+    end
+
+    it "can't assign name to permanent module" do
+      -> { Object.set_temporary_name("Foo") }.should raise_error(RuntimeError, "can't change permanent name")
+    end
+
+    it "can assign a temporary name to a nested module" do
+      m = Module.new
+      module m::N; end
+      m::N.name.should =~ /\A#<Module:0x\h+>::N\z/
+
+      m::N.set_temporary_name("Foo")
+      m::N.name.should == "Foo"
+
+      m::N.set_temporary_name(nil)
+      m::N.name.should be_nil
```

Comment:
I find it strange that this does not revert to the original temporary name at the top of this test case. It seems like if you set  an explicit temporary name, there is no way to go back to the original internal temporary name, even if you wanted to. I am not sure if that's very desirable.

Reply:
You can get it back to the "original" name by assigning it to a constant again. That's what constant assignment does, i.e. to get the original name, you'd need to `m::N = m::N`. That's why it has its initial name, because there is an implicit constant assignment in `module m::N; end`. I don't think `set_temporary_name` should attempt to reproduce that (it might even be impossible), e.g.:

```ruby
m = Module.new
# => #<Module:0x000000010662d160>

n = Module.new
# => #<Module:0x00000001065a5a08>

m::N = n
# => #<Module:0x000000010662d160>::N
m::M = n
# => #<Module:0x000000010662d160>::N

n.set_temporary_name(nil)

# what should be the name?
# #<Module:0x000000010662d160>::N
# or
# #<Module:0x000000010662d160>::M
```

This problem is only resolved by constant assignment.

Reply:
Right, I see. Thanks for the explanation. If there is such an interplay between setting temporary name and assigning to a constant name again, do you think it would be a good idea to add a spec to make it explicit that is the case?

Reply:
Okay, I've added a spec for what I think you are referring to.